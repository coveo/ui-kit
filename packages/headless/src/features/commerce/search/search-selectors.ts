import {createSelector} from '@reduxjs/toolkit';
import {CommerceEngineState} from '../../../app/commerce-engine/commerce-engine';

export const responseIdSelector = createSelector(
  (state: CommerceEngineState) => state.commerceSearch.responseId,
  (responseId) => responseId
);
