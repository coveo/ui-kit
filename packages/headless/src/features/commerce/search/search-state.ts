import {CommerceAPIErrorStatusResponse} from '../../../api/commerce/commerce-api-error-response';
import {Product} from '../../../api/commerce/common/product';
import {AnyFacetResponse} from '../facets/facet-set/interfaces/response';

export interface CommerceSearchState {
  error: CommerceAPIErrorStatusResponse | null;
  isLoading: boolean;
  requestId: string;
  responseId: string;
  products: Product[];
  facets: AnyFacetResponse[];
  queryExecuted: string;
}

export const getCommerceSearchInitialState = (): CommerceSearchState => ({
  error: null,
  isLoading: false,
  requestId: '',
  responseId: '',
  products: [],
  facets: [],
  queryExecuted: '',
});
