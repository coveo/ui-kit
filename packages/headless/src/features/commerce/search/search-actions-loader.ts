import {AsyncThunkAction, PayloadAction} from '@reduxjs/toolkit';
import {AsyncThunkCommerceOptions} from '../../../api/commerce/commerce-api-client';
import {CommerceEngine} from '../../../app/commerce-engine/commerce-engine';
import {
  PrepareForSearchWithQueryPayload,
  PromoteChildToParentPayload,
  QuerySearchCommerceAPIThunkReturn,
  executeSearch,
  fetchMoreProducts,
  prepareForSearchWithQuery,
  promoteChildToParent,
} from './search-actions';
import {StateNeededByExecuteSearch} from './search-actions-thunk-processor';
import {commerceSearchReducer as commerceSearch} from './search-slice';

export type {PrepareForSearchWithQueryPayload};

/**
 * The search action creators.
 */
export interface SearchActionCreators {
  /**
   * Executes a search query.
   *
   * @returns A dispatchable action.
   */
  executeSearch(): AsyncThunkAction<
    QuerySearchCommerceAPIThunkReturn,
    void,
    AsyncThunkCommerceOptions<StateNeededByExecuteSearch>
  >;

  /**
   * Fetches and additional page of products and appends it to the current list.
   *
   * @returns A dispatchable action.
   */
  fetchMoreProducts(): AsyncThunkAction<
    QuerySearchCommerceAPIThunkReturn | null,
    void,
    AsyncThunkCommerceOptions<StateNeededByExecuteSearch>
  >;

  /**
   * Updates the query, resets the pagination, and optionally clears all facets in preparation for a new search query.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  prepareForSearchWithQuery(
    payload: PrepareForSearchWithQueryPayload
  ): AsyncThunkAction<
    void,
    PrepareForSearchWithQueryPayload,
    AsyncThunkCommerceOptions<StateNeededByExecuteSearch>
  >;

  /**
   * Promotes a child product to a parent product.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  promoteChildToParent(
    payload: PromoteChildToParentPayload
  ): PayloadAction<PromoteChildToParentPayload>;
}

/**
 * Loads the commerce search reducer and returns the available search action creators.
 *
 * @param engine - The headless commerce engine.
 * @returns An object holding the search action creators.
 */
export function loadSearchActions(
  engine: CommerceEngine
): SearchActionCreators {
  engine.addReducers({commerceSearch});

  return {
    executeSearch,
    fetchMoreProducts,
    prepareForSearchWithQuery,
    promoteChildToParent,
  };
}
