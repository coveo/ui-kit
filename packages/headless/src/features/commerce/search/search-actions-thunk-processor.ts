import {isNullOrUndefined} from '@coveo/bueno';
import {UnknownAction} from '@reduxjs/toolkit';
import {ThunkDispatch} from 'redux-thunk';
import {
  CommerceAPIClient,
  CommerceAPIResponse,
  isErrorResponse,
} from '../../../api/commerce/commerce-api-client';
import {CommerceAPIErrorStatusResponse} from '../../../api/commerce/commerce-api-error-response';
import {CommerceAPIRequest} from '../../../api/commerce/common/request';
import {CommerceSearchRequest} from '../../../api/commerce/search/request';
import {SearchCommerceSuccessResponse} from '../../../api/commerce/search/response';
import {ClientThunkExtraArguments} from '../../../app/thunk-extra-arguments';
import {
  CommerceSearchSection,
  CommerceQuerySection,
  CommerceDidYouMeanSection,
  TriggerSection,
} from '../../../state/state-sections';
import {
  ListingAndSearchStateNeededByQueryCommerceAPI,
  buildCommerceAPIRequest,
} from '../common/actions';
import {updateQuery} from '../query/query-actions';
import {
  applyQueryTriggerModification,
  updateIgnoreQueryTrigger,
} from '../triggers/triggers-actions';
import {QuerySearchCommerceAPIThunkReturn} from './search-actions';
import {queryExecutedFromResponseSelector} from './search-selectors';

interface FetchedResponse {
  response: CommerceAPIResponse<SearchCommerceSuccessResponse>;
  duration: number;
  queryExecuted: string;
  requestExecuted: CommerceAPIRequest;
}

export type StateNeededByExecuteSearch =
  ListingAndSearchStateNeededByQueryCommerceAPI &
    CommerceSearchSection &
    CommerceQuerySection &
    CommerceDidYouMeanSection &
    TriggerSection;

export interface AsyncThunkConfig {
  getState: () => StateNeededByExecuteSearch;
  dispatch: ThunkDispatch<
    StateNeededByExecuteSearch,
    ClientThunkExtraArguments<CommerceAPIClient> & {
      apiClient?: CommerceAPIClient | undefined;
    },
    UnknownAction
  >;

  rejectWithValue: (err: CommerceAPIErrorStatusResponse) => unknown;
  extra: ClientThunkExtraArguments<CommerceAPIClient>;
}

type ValidReturnTypeFromProcessingStep<RejectionType> =
  | QuerySearchCommerceAPIThunkReturn
  | RejectionType;

export class AsyncSearchThunkProcessor<RejectionType> {
  constructor(private config: AsyncThunkConfig) {}

  public async process(
    fetched: FetchedResponse
  ): Promise<ValidReturnTypeFromProcessingStep<RejectionType>> {
    return (
      this.processQueryErrorOrContinue(fetched) ??
      (await this.processQueryCorrectionsOrContinue(fetched)) ??
      (await this.processQueryTriggersOrContinue(fetched)) ??
      this.processSuccessResponse(fetched)
    );
  }

  public async fetchFromAPI(
    request: CommerceAPIRequest | CommerceSearchRequest
  ) {
    const startedAt = new Date().getTime();
    const response = await this.extra.apiClient.search(request);
    const duration = new Date().getTime() - startedAt;
    const queryExecuted = this.getState().commerceQuery.query || '';
    return {response, duration, queryExecuted, requestExecuted: request};
  }

  private processSuccessResponse(
    fetched: FetchedResponse
  ): QuerySearchCommerceAPIThunkReturn {
    return {
      ...fetched,
      response: this.getSuccessResponse(fetched)!,
      originalQuery: this.getCurrentQuery(),
    };
  }

  private processQueryErrorOrContinue(
    fetched: FetchedResponse
  ): ValidReturnTypeFromProcessingStep<RejectionType> | null {
    if (isErrorResponse(fetched.response)) {
      return this.rejectWithValue(fetched.response.error) as RejectionType;
    }

    return null;
  }

  private async processQueryCorrectionsOrContinue(
    fetched: FetchedResponse
  ): Promise<ValidReturnTypeFromProcessingStep<RejectionType> | null> {
    const state = this.getState();
    const successResponse = this.getSuccessResponse(fetched);

    if (!successResponse || !state.didYouMean) {
      return null;
    }
    const {queryCorrection} = successResponse;

    const shouldExecuteQueryCorrection =
      !isNullOrUndefined(queryCorrection) &&
      !isNullOrUndefined(queryCorrection.correctedQuery);

    if (!shouldExecuteQueryCorrection) {
      return null;
    }

    const {correctedQuery, originalQuery} = successResponse.queryCorrection!;

    this.onUpdateQueryForCorrection(correctedQuery);

    return {
      ...fetched,
      response: {
        ...successResponse,
      },
      queryExecuted: queryExecutedFromResponseSelector(state, successResponse),
      originalQuery,
    };
  }

  private async processQueryTriggersOrContinue(
    fetched: FetchedResponse
  ): Promise<ValidReturnTypeFromProcessingStep<RejectionType> | null> {
    const successResponse = this.getSuccessResponse(fetched);
    if (!successResponse) {
      return null;
    }
    const correctedQuery =
      (successResponse.triggers.find((trigger) => trigger.type === 'query')
        ?.content as string) || '';

    if (!correctedQuery) {
      return null;
    }

    const ignored = this.getState().triggers?.queryModification.queryToIgnore;

    if (ignored === correctedQuery) {
      this.dispatch(updateIgnoreQueryTrigger({q: ''}));
      return null;
    }

    const originalQuery = this.getCurrentQuery();
    const retried =
      await this.automaticallyRetryQueryWithTriggerModification(correctedQuery);

    if (isErrorResponse(retried.response)) {
      return this.rejectWithValue(retried.response.error) as RejectionType;
    }

    return {
      ...retried,
      response: {
        ...retried.response.success,
      },
      originalQuery,
    };
  }

  private async automaticallyRetryQueryWithTriggerModification(
    modified: string
  ) {
    this.dispatch(
      applyQueryTriggerModification({
        newQuery: modified,
        originalQuery: this.getCurrentQuery(),
      })
    );
    this.onUpdateQueryForCorrection(modified);
    const fetched = await this.fetchFromAPI({
      ...buildCommerceAPIRequest(this.getState(), this.navigatorContext),
      query: modified,
    });

    return fetched;
  }

  private get dispatch() {
    return this.config.dispatch;
  }

  private get rejectWithValue() {
    return this.config.rejectWithValue;
  }

  private getState() {
    return this.config.getState();
  }

  private get navigatorContext() {
    return this.config.extra.navigatorContext;
  }

  private getCurrentQuery() {
    const state = this.getState();
    return state.commerceQuery.query !== undefined
      ? state.commerceQuery.query
      : '';
  }

  private getSuccessResponse(fetched: FetchedResponse) {
    if (isErrorResponse(fetched.response)) {
      return null;
    }
    return fetched.response.success;
  }

  private get extra() {
    return this.config.extra;
  }

  private onUpdateQueryForCorrection(modification: string) {
    this.dispatch(updateQuery({query: modification}));
  }
}
