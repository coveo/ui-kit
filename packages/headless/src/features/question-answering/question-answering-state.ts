import {QuestionAnswerDocumentIdentifier} from '../../api/search/search/question-answering';

export interface QuestionAnsweringRelatedQuestionState
  extends QuestionAnswerDocumentIdentifier {
  expanded: boolean;
}
export interface QuestionAnsweringState {
  /**
   * Determines if the snippet is liked, or upvoted by the end user.
   */
  liked: boolean;
  /**
   * Determines if the snippet is disliked, or downvoted by the end user.
   */
  disliked: boolean;
  /**
   * Determines if the snippet is expanded.
   */
  expanded: boolean;
  relatedQuestions: QuestionAnsweringRelatedQuestionState[];
}

export const getQuestionAnsweringInitialState: () => QuestionAnsweringState =
  () => ({
    liked: false,
    disliked: false,
    expanded: false,
    relatedQuestions: [],
  });
