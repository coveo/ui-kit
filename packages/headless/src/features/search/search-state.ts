import {SearchAPIErrorWithStatusCode} from '../../api/search/search-api-error-response';
import {Result} from '../../api/search/search/result';
import {SearchResponseSuccess} from '../../api/search/search/search-response';

export interface SearchState {
  /** The search response. For a full description, refer to {@link https://docs.coveo.com/en/13/cloud-v2-api-reference/search-api#operation/searchUsingPost}*/
  response: SearchResponseSuccess;
  /**
   * The time it took to execute the request against the Coveo platform.
   */
  duration: number;
  /**
   * The query (searchbox content) that was executed against the Coveo platform.
   */
  queryExecuted: string;
  /**
   * The error returned by the Coveo platform while executing the search request, if any. `null` otherwise.
   */
  error: SearchAPIErrorWithStatusCode | null;
  /**
   * Specifies if the query was automatically corrected by Headless.
   * This happens when there is no result returned by the API for a particular mispelling.
   */
  automaticallyCorrected: boolean;
  /**
   * `true` if the search request is currently being executed against the Coveo platform, `false` otherwise.
   */
  isLoading: boolean;
  /**
   * The list of results.
   */
  results: Result[];
  /**
   * The unique ID of the query, taken from the response's `searchUid. This value will not change when loading more results.
   */
  searchResponseId: string;
}

export function emptyQuestionAnswer() {
  return {
    answerSnippet: '',
    documentId: {
      contentIdKey: '',
      contentIdValue: '',
    },
    question: '',
    relatedQuestions: [],
    score: 0,
  };
}

export function getSearchInitialState(): SearchState {
  return {
    response: {
      results: [],
      searchUid: '',
      totalCountFiltered: 0,
      facets: [],
      queryCorrections: [],
      triggers: [],
      questionAnswer: emptyQuestionAnswer(),
    },
    duration: 0,
    queryExecuted: '',
    error: null,
    automaticallyCorrected: false,
    isLoading: false,
    results: [],
    searchResponseId: '',
  };
}
