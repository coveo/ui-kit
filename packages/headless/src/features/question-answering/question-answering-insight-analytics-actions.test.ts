import {
  buildMockInsightEngine,
  MockInsightEngine,
} from '../../test/mock-engine';
import {buildMockInsightState} from '../../test/mock-insight-state';
import {buildMockRaw} from '../../test/mock-raw';
import {buildMockResult} from '../../test/mock-result';
import {buildMockSearchState} from '../../test/mock-search-state';
import {emptyQuestionAnswer} from '../search/search-state';
import {
  logExpandSmartSnippet,
  logExpandSmartSnippetSuggestion,
  logCollapseSmartSnippetSuggestion,
  logOpenSmartSnippetSuggestionSource,
  logOpenSmartSnippetSuggestionInlineLink,
  logCollapseSmartSnippet,
  logLikeSmartSnippet,
  logDislikeSmartSnippet,
  logOpenSmartSnippetSource,
  logOpenSmartSnippetInlineLink,
  logOpenSmartSnippetFeedbackModal,
  logCloseSmartSnippetFeedbackModal,
  logSmartSnippetFeedback,
  logSmartSnippetDetailedFeedback,
} from './question-answering-insight-analytics-actions';
import {getQuestionAnsweringInitialState} from './question-answering-state';

const mockLogExpandSmartSnippet = jest.fn();
const mockLogCollapseSmartSnippet = jest.fn();
const mockLogLikeSmartSnippet = jest.fn();
const mockLogDislikeSmartSnippet = jest.fn();
const mockLogOpenSmartSnippetSource = jest.fn();
const mockLogOpenSmartSnippetInlineLink = jest.fn();
const mockLogOpenSmartSnippetFeedbackModal = jest.fn();
const mockLogCloseSmartSnippetFeedbackModal = jest.fn();
const mockLogSmartSnippetFeedback = jest.fn();
const mockLogSmartSnippetDetailedFeedback = jest.fn();
const mockLogExpandSmartSnippetSuggestion = jest.fn();
const mockLogCollapseSmartSnippetSuggestion = jest.fn();
const mockLogOpenSmartSnippetSuggestionSource = jest.fn();
const mockLogOpenSmartSnippetSuggestionInlineLink = jest.fn();

jest.mock('coveo.analytics', () => {
  const mockCoveoInsightClient = jest.fn(() => ({
    disable: jest.fn(),
    logExpandSmartSnippet: mockLogExpandSmartSnippet,
    logCollapseSmartSnippet: mockLogCollapseSmartSnippet,
    logLikeSmartSnippet: mockLogLikeSmartSnippet,
    logDislikeSmartSnippet: mockLogDislikeSmartSnippet,
    logOpenSmartSnippetSource: mockLogOpenSmartSnippetSource,
    logOpenSmartSnippetInlineLink: mockLogOpenSmartSnippetInlineLink,
    logOpenSmartSnippetFeedbackModal: mockLogOpenSmartSnippetFeedbackModal,
    logCloseSmartSnippetFeedbackModal: mockLogCloseSmartSnippetFeedbackModal,
    logSmartSnippetFeedbackReason: mockLogSmartSnippetFeedback,
    logSmartSnippetDetailedFeedback: mockLogSmartSnippetDetailedFeedback,
    logExpandSmartSnippetSuggestion: mockLogExpandSmartSnippetSuggestion,
    logCollapseSmartSnippetSuggestion: mockLogCollapseSmartSnippetSuggestion,
    logOpenSmartSnippetSuggestionSource:
      mockLogOpenSmartSnippetSuggestionSource,
    logOpenSmartSnippetSuggestionInlineLink:
      mockLogOpenSmartSnippetSuggestionInlineLink,
  }));

  return {
    CoveoInsightClient: mockCoveoInsightClient,
    history: {HistoryStore: jest.fn()},
  };
});

const exampleQuestion = 'Where am I?';
const exampleAnswerSnippet = 'You are here.';
const exampleScore = 9001;
const exampleSubject = 'example subject';
const exampleDescription = 'example description';
const exampleCaseId = '1234';
const exampleCaseNumber = '5678';
const examplePermanentId = 'example permanent id';
const exampleQuestionAnswerId = '1';
const exampleFeedback = 'does_not_answer';
const exampleFeedbackDetails = 'example feedback details';

const exampleRelatedQuestion = {
  question: 'example question',
  answerSnippet: 'example answer',
  documentId: {
    contentIdKey: 'permanentid',
    contentIdValue: examplePermanentId,
  },
  score: 1,
};

const exampleQuestionAnswer = {
  question: exampleQuestion,
  answerSnippet: exampleAnswerSnippet,
  documentId: {
    contentIdKey: 'permanentid',
    contentIdValue: examplePermanentId,
  },
  score: exampleScore,
};

const exampleQuestionAnsweringState = {
  liked: false,
  disliked: false,
  expanded: false,
  feedbackModalOpen: false,
  questionAnsweringId: 1,
};

const exampleRelatedQuestionState = {
  expanded: false,
  questionAnswerId: exampleQuestionAnswerId,
  contentIdKey: 'permanentid',
  contentIdValue: examplePermanentId,
};

const expectedCaseContext = {
  caseContext: {
    Case_Subject: exampleSubject,
    Case_Description: exampleDescription,
  },
  caseId: exampleCaseId,
  caseNumber: exampleCaseNumber,
};

const expectedDocumentInfo = {
  queryPipeline: '',
  documentUri: 'example documentUri',
  documentUriHash: 'example documentUriHash',
  collectionName: 'example collectionName',
  sourceName: 'example sourceName',
  documentPosition: 1,
  documentTitle: 'example documentTitle',
  documentUrl: 'example documentUrl',
  rankingModifier: 'example rankingModifier',
  documentAuthor: 'unknown',
};

const resultParams = {
  title: 'example documentTitle',
  uri: 'example documentUri',
  printableUri: 'printable-uri',
  clickUri: 'example documentUrl',
  rankingModifier: 'example rankingModifier',
  raw: buildMockRaw({
    urihash: 'example documentUriHash',
    source: 'example sourceName',
    collection: 'example collectionName',
    permanentid: examplePermanentId,
  }),
};
const exampleResult = buildMockResult(resultParams);

const exampleInlineLink = {
  linkText: 'example link text',
  linkURL: 'example link url',
};

describe('the analytics related to the question answering feature in the insight use case', () => {
  let engine: MockInsightEngine;

  beforeEach(() => {
    engine = buildMockInsightEngine({
      state: buildMockInsightState({
        insightCaseContext: {
          caseContext: {
            Case_Subject: exampleSubject,
            Case_Description: exampleDescription,
          },
          caseId: exampleCaseId,
          caseNumber: exampleCaseNumber,
        },
        questionAnswering: {
          ...getQuestionAnsweringInitialState(),
          ...exampleQuestionAnsweringState,
          relatedQuestions: [exampleRelatedQuestionState],
        },
        search: buildMockSearchState({
          results: [exampleResult],
          questionAnswer: {
            ...emptyQuestionAnswer(),
            ...exampleQuestionAnswer,
            relatedQuestions: [exampleRelatedQuestion],
          },
        }),
      }),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should log #logExpandSmartSnippet with the case payload', async () => {
    await engine.dispatch(logExpandSmartSnippet());

    const mockToUse = mockLogExpandSmartSnippet;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logCollapseSmartSnippet with the case payload', async () => {
    await engine.dispatch(logCollapseSmartSnippet());

    const mockToUse = mockLogCollapseSmartSnippet;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logLikeSmartSnippet with the case payload', async () => {
    await engine.dispatch(logLikeSmartSnippet());

    const mockToUse = mockLogLikeSmartSnippet;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logDislikeSmartSnippet with the case payload', async () => {
    await engine.dispatch(logDislikeSmartSnippet());

    const mockToUse = mockLogDislikeSmartSnippet;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logOpenSmartSnippetSource with the right payload', async () => {
    await engine.dispatch(logOpenSmartSnippetSource());

    const mockToUse = mockLogOpenSmartSnippetSource;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
    expect(mockToUse.mock.calls[0][1]).toStrictEqual({
      contentIDKey: exampleQuestionAnswer.documentId.contentIdKey,
      contentIDValue: exampleQuestionAnswer.documentId.contentIdValue,
    });
    expect(mockToUse.mock.calls[0][2]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logOpenSmartSnippetInlineLink with the right payload', async () => {
    await engine.dispatch(logOpenSmartSnippetInlineLink(exampleInlineLink));

    const mockToUse = mockLogOpenSmartSnippetInlineLink;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
    expect(mockToUse.mock.calls[0][1]).toStrictEqual({
      contentIDKey: exampleQuestionAnswer.documentId.contentIdKey,
      contentIDValue: exampleQuestionAnswer.documentId.contentIdValue,
      ...exampleInlineLink,
    });
    expect(mockToUse.mock.calls[0][2]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logOpenSmartSnippetFeedbackModal with the case payload', async () => {
    await engine.dispatch(logOpenSmartSnippetFeedbackModal());

    const mockToUse = mockLogOpenSmartSnippetFeedbackModal;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logCloseSmartSnippetFeedbackModal with the case payload', async () => {
    await engine.dispatch(logCloseSmartSnippetFeedbackModal());

    const mockToUse = mockLogCloseSmartSnippetFeedbackModal;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logSmartSnippetFeedback with the right payload', async () => {
    await engine.dispatch(logSmartSnippetFeedback(exampleFeedback));

    const mockToUse = mockLogSmartSnippetFeedback;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(exampleFeedback);
    expect(mockToUse.mock.calls[0][1]).toStrictEqual(undefined);
    expect(mockToUse.mock.calls[0][2]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logSmartSnippetDetailedFeedback with the right payload', async () => {
    await engine.dispatch(
      logSmartSnippetDetailedFeedback(exampleFeedbackDetails)
    );

    const expectedFeedbackReason = 'other';

    const mockToUse = mockLogSmartSnippetFeedback;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedFeedbackReason);
    expect(mockToUse.mock.calls[0][1]).toStrictEqual(exampleFeedbackDetails);
    expect(mockToUse.mock.calls[0][2]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logExpandSmartSnippetSuggestion with the right payload', async () => {
    await engine.dispatch(
      logExpandSmartSnippetSuggestion({
        questionAnswerId: exampleQuestionAnswerId,
      })
    );
    const expectedRelatedQuestionPayload = {
      question: exampleRelatedQuestion.question,
      answerSnippet: exampleRelatedQuestion.answerSnippet,
      documentId: exampleRelatedQuestion.documentId,
    };

    const mockToUse = mockLogExpandSmartSnippetSuggestion;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(
      expectedRelatedQuestionPayload
    );
    expect(mockToUse.mock.calls[0][1]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logCollapseSmartSnippetSuggestion with the right payload', async () => {
    await engine.dispatch(
      logCollapseSmartSnippetSuggestion({
        questionAnswerId: exampleQuestionAnswerId,
      })
    );

    const expectedRelatedQuestionPayload = {
      question: exampleRelatedQuestion.question,
      answerSnippet: exampleRelatedQuestion.answerSnippet,
      documentId: exampleRelatedQuestion.documentId,
    };

    const mockToUse = mockLogCollapseSmartSnippetSuggestion;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(
      expectedRelatedQuestionPayload
    );
    expect(mockToUse.mock.calls[0][1]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logOpenSmartSnippetSuggestionSource with the right payload', async () => {
    await engine.dispatch(
      logOpenSmartSnippetSuggestionSource({
        questionAnswerId: exampleQuestionAnswerId,
      })
    );

    const expectedRelatedQuestionPayload = {
      question: exampleRelatedQuestion.question,
      answerSnippet: exampleRelatedQuestion.answerSnippet,
      documentId: exampleRelatedQuestion.documentId,
    };

    const mockToUse = mockLogOpenSmartSnippetSuggestionSource;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
    expect(mockToUse.mock.calls[0][1]).toStrictEqual(
      expectedRelatedQuestionPayload
    );
    expect(mockToUse.mock.calls[0][2]).toStrictEqual(expectedCaseContext);
  });

  it('should log #logOpenSmartSnippetSuggestionInlineLink with the right payload', async () => {
    await engine.dispatch(
      logOpenSmartSnippetSuggestionInlineLink(
        {questionAnswerId: exampleQuestionAnswerId},
        exampleInlineLink
      )
    );

    const expectedRelatedQuestionPayload = {
      question: exampleRelatedQuestion.question,
      answerSnippet: exampleRelatedQuestion.answerSnippet,
      documentId: exampleRelatedQuestion.documentId,
    };

    const mockToUse = mockLogOpenSmartSnippetSuggestionInlineLink;
    expect(mockToUse).toBeCalledTimes(1);
    expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
    expect(mockToUse.mock.calls[0][1]).toStrictEqual({
      ...expectedRelatedQuestionPayload,
      ...exampleInlineLink,
    });
    expect(mockToUse.mock.calls[0][2]).toStrictEqual(expectedCaseContext);
  });
});
