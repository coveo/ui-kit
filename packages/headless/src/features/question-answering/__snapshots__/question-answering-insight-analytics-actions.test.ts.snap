// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logCollapseSmartSnippet with the case payload 1`] = `
[
  "SmartSnippets.AnswerAction",
  {
    "action": "collapse",
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logCollapseSmartSnippetSuggestion with the right payload 1`] = `
[
  "SmartSnippets.AnswerAction",
  {
    "action": "collapse",
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippetSuggestion",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logDislikeSmartSnippet with the case payload 1`] = `
[
  "SmartSnippets.AnswerAction",
  {
    "action": "dislike",
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logExpandSmartSnippet with the case payload 1`] = `
[
  "SmartSnippets.AnswerAction",
  {
    "action": "expand",
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logExpandSmartSnippetSuggestion with the right payload 1`] = `
[
  "SmartSnippets.AnswerAction",
  {
    "action": "expand",
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippetSuggestion",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logLikeSmartSnippet with the case payload 1`] = `
[
  "SmartSnippets.AnswerAction",
  {
    "action": "like",
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logOpenSmartSnippetInlineLink with the right payload 1`] = `
[
  "SmartSnippets.SourceClick",
  {
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logOpenSmartSnippetSource with the right payload 1`] = `
[
  "SmartSnippets.SourceClick",
  {
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logOpenSmartSnippetSuggestionInlineLink with the right payload 1`] = `
[
  "SmartSnippets.SourceClick",
  {
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippetSuggestion",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logOpenSmartSnippetSuggestionSource with the right payload 1`] = `
[
  "SmartSnippets.SourceClick",
  {
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "responseId": "456",
    "snippetType": "SmartSnippetSuggestion",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logSmartSnippetDetailedFeedback with the right payload 1`] = `
[
  "SmartSnippets.SubmitFeedback",
  {
    "additionalNotes": "example feedback details",
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "reason": "other",
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;

exports[`question answering insight analytics actions > when analyticsMode is \`next\` > should log #logSmartSnippetFeedback with the right payload 1`] = `
[
  "SmartSnippets.SubmitFeedback",
  {
    "itemMetadata": {
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example permanent id",
      "url": "example documentUrl",
    },
    "reason": "doesNotAnswer",
    "responseId": "456",
    "snippetType": "SmartSnippet",
  },
]
`;
